src/lib/completion.ts

```diff
@@ -1,131 +1,216 @@
+		if (content.includes('Final Answer:')) {
+			if (this.verbose) {
+				console.log(colors.fg.red, 'Model response: ', colors.style.reset);
+				console.log(colors.fg.yellow, content, colors.style.reset);
 			}
-		});
+			this.chain.push(JSON.stringify({ chatbot: content }));
+			const parts = content.split('Final Answer:');
+			const input = parts[parts.length - 1].trim();
+			const finalAnswers = { output: input };
+			return { log: content, returnValues: finalAnswers };
+		}
+		const match = /Action: (.*)\nAction Input: (.*)/s.exec(content);
+		if (match == null) {
+			// this.chain.slice(-1, 1);
+			this.chain.push(JSON.stringify({ chatbot: content }));
+			console.warn('Could not parse output');
+			console.warn(content);
+			const finalAnswers = { output: content };
+			return { log: content, returnValues: finalAnswers };
+		}
+		if (this.verbose) {
+			console.log(colors.fg.red, 'Model response: ', colors.style.reset);
+			console.log(colors.fg.yellow, content, colors.style.reset);
+		}
+		this.chain.push(JSON.stringify({ chatbot: content }));
+		return {
+			tool: match[1].trim(),
+			toolInput: match[2].trim().replace(/^"+|"+$/g, ''),
+			log: content
+		};
 	}
 
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	public async query(history: ChatHistory[]): Promise<any> {
-		const vector_response = await this.get_vector_response(history[history.length - 1].content);
-
-		const context = vector_response.map((content: string) => {
-			return new SystemMessage({ content });
-		});
-
-		if (context.length == 0) {
-			context.push(
-				new SystemMessage({
-					content:
-						'You dont have data on this content, you may want to respond with "sorry I cannot answer that as I do not have enough information"'
-				})
-			);
+	public async query(history: ChatHistory[], input: any): Promise<any> {
+		if (this.executor == undefined) {
+			console.warn("ChatbotCompletion's setup was not called, calling setup");
+			this.setup();
+			if (this.executor == undefined) {
+				console.warn('setup failed');
+				return 'Error';
+			}
 		}
-
-		const chat_history = [
-			new SystemMessage({
-				content: `You are a kind, professional, understanding, and enthusiastic
-                    assistant that is an expert in mechanical, electrical, and software engineering, but most importantly FIRST robotics
-                    and helping all levels of frc robotics teams. Avoiding repeating the same information and useless statements.
-                    The current date is ${new Date()}.`
-			}),
-			...context,
-			...this.generate_history(history)
-		];
-
-		// TODO(max): Add history pruning based on token length
-
-		return await this.model.pipe(new BytesOutputParser()).stream(chat_history);
+		if (this.do_history) this.history = history;
+		const result = await this.executor.invoke({ input: input });
+		if (this.generate_data)
+			fs.writeFile('data/logs/' + input + '.jsonl', this.chain.join('\n'), (err) => {
+				if (err) throw err;
```

Comment:
I don't like us throwing an error here. I have a feeling its good enough to put a console warn and move on as this isn't 100% necessary for continuing to provide output to the user.